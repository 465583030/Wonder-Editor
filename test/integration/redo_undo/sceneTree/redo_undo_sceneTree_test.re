open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: sceneTree", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => sandbox := createSandbox());
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    let _simulateTwiceDragEvent = () => {
      let firstCameraDomIndex =
        SceneTreeNodeDomTool.OperateDefaultScene.getFirstCameraDomIndex();
      let firstCubeDomIndex =
        SceneTreeNodeDomTool.OperateDefaultScene.getFirstCubeDomIndex();
      let secondCubeDomIndex =
        SceneTreeNodeDomTool.OperateDefaultScene.getSecondCubeDomIndex();

      let component =
        BuildComponentTool.buildSceneTree(
          TestTool.buildAppStateSceneGraphFromEngine(),
        );
      BaseEventTool.triggerComponentEvent(
        component,
        SceneTreeEventTool.triggerDragStart(secondCubeDomIndex),
      );
      BaseEventTool.triggerComponentEvent(
        component,
        SceneTreeEventTool.triggerDragEnter(firstCameraDomIndex),
      );
      BaseEventTool.triggerComponentEvent(
        component,
        SceneTreeEventTool.triggerDragDrop(firstCameraDomIndex),
      );

      let component2 =
        BuildComponentTool.buildSceneTree(
          TestTool.buildAppStateSceneGraphFromEngine(),
        );
      BaseEventTool.triggerComponentEvent(
        component2,
        SceneTreeEventTool.triggerDragStart(firstCubeDomIndex),
      );
      BaseEventTool.triggerComponentEvent(
        component2,
        SceneTreeEventTool.triggerDragEnter(firstCameraDomIndex),
      );
      BaseEventTool.triggerComponentEvent(
        component2,
        SceneTreeEventTool.triggerDragDrop(firstCameraDomIndex),
      );
    };
    let _beforeEach = () => {
      MainEditorSceneTool.initState(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorSceneTool.setFirstBoxToBeCurrentSceneTreeNode,
      );
    };

    RedoUndoTool.testRedoUndoTwoStep(
      sandbox,
      "get scene tree from engine",
      (_simulateTwiceDragEvent, _beforeEach, () => ()),
      BuildComponentForCurryTool.buildSceneTree,
    );

    describe("fix bug", () => {
      let execChangeMaterialColorWork = (currentGameObjectMaterial, newColor) =>
        PickColorEventTool.triggerChangeLightMaterialColor(
          currentGameObjectMaterial,
          newColor,
        );

      let execChangeTransformWork = () => {
        let currentGameObjectTransform =
          GameObjectTool.getCurrentSceneTreeNodeTransform();
        let transformComponent =
          BuildComponentTool.buildMainEditorTransformComponent(
            TestTool.buildEmptyAppState(),
            currentGameObjectTransform,
          );
        BaseEventTool.triggerComponentEvent(
          transformComponent,
          TransformEventTool.triggerChangePositionX("11.25"),
        );
        BaseEventTool.triggerComponentEvent(
          transformComponent,
          TransformEventTool.triggerBlurPositionX("11.25"),
        );
      };

      beforeEach(() => {
        MainEditorSceneTool.initState(~sandbox, ());
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxToBeCurrentSceneTreeNode,
        );
      });
      test(
        "the workflow:
        click treeNote set currentSceneTreeNode;
        change material color;
        change transform x value;
        click undo, engineState is error",
        () => {
          let currentGameObjectMaterial =
            GameObjectTool.getCurrentGameObjectLightMaterial();
          let newColor = {
            "hex": "#7df1e8",
            "rgb": {
              "r": 125,
              "g": 241,
              "b": 232,
            },
          };

          execChangeMaterialColorWork(currentGameObjectMaterial, newColor);
          execChangeTransformWork();

          StateHistoryToolEditor.undo();

          LightMaterialEngineService.getLightMaterialDiffuseColor(
            currentGameObjectMaterial,
          )
          |> StateLogicService.getEngineStateToGetData
          |> Color.getHexString
          |> expect == newColor##hex;
        },
      );
    });
  });