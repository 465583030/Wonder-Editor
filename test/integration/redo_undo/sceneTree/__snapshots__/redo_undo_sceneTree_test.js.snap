// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo one step undo step which from second to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        draggable={true}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "border": "0px",
          }
        }
      >
        <div
          className="item-triangle"
          onClick={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        Scene
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
            onClick={[Function]}
          >
            <img
              src="./public/img/right.png"
            />
          </div>
          Camera
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "background": "#5C7EA6",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Direction Light
        </li>
      </ul>
    </ul>
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo two step undo step which from second to zero,redo step which from zero to second 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        draggable={true}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "border": "0px",
          }
        }
      >
        <div
          className="item-triangle"
          onClick={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        Scene
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
            onClick={[Function]}
          >
            <img
              src="./public/img/right.png"
            />
          </div>
          Camera
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Direction Light
        </li>
      </ul>
    </ul>
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test not undo 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        draggable={true}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "border": "0px",
          }
        }
      >
        <div
          className="item-triangle"
          onClick={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        Scene
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
            onClick={[Function]}
          >
            <img
              src="./public/img/right.png"
            />
          </div>
          Camera
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Direction Light
        </li>
      </ul>
    </ul>
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test undo one step step which from second to first 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        draggable={true}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "border": "0px",
          }
        }
      >
        <div
          className="item-triangle"
          onClick={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        Scene
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
            onClick={[Function]}
          >
            <img
              src="./public/img/right.png"
            />
          </div>
          Camera
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "background": "#5C7EA6",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Direction Light
        </li>
      </ul>
    </ul>
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test undo two step step which from second to zero 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        draggable={true}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "border": "0px",
          }
        }
      >
        <div
          className="item-triangle"
          onClick={[Function]}
        >
          <img
            src="./public/img/down.png"
          />
        </div>
        Scene
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Camera
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "background": "#5C7EA6",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          draggable={true}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "border": "0px",
            }
          }
        >
          <div
            className="item-triangle"
          />
          Direction Light
        </li>
      </ul>
    </ul>
  </article>
</article>
`;
